File: Reference/src/Brontide.Reference.Studio/ProviderPolicyAuthorityRotationDistribution.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Brontide.Reference.Studio;

public sealed record ProviderPolicyAuthorityRotationDistributionRequest(
    string Challenge,
    long PolicySequence,
    ProviderPublisherTrustPolicyId? PolicyIdentity,
    long AuthorityGeneration,
    ProviderPublisherTrustPolicyAuthorityId ActiveAuthority);

public sealed record ProviderPolicyAuthorityRotationDistributionResponse(
    string Challenge,
    long PolicySequence,
    ProviderPublisherTrustPolicyId? PolicyIdentity,
    long AuthorityGeneration,
    ProviderPublisherTrustPolicyAuthorityId ActiveAuthority,
    long IssuedAtUnixSeconds,
    long ExpiresAtUnixSeconds,
    ProviderPolicyAuthorityRotationStatement? Rotation,
    string Algorithm,
    string EndpointPublicKeySpkiBase64,
    string SignatureBase64);

public interface IProviderPolicyAuthorityRotationDistributionSource
{
    Task<ProviderPolicyAuthorityRotationDistributionResponse> FetchAsync(
        ProviderPolicyAuthorityRotationDistributionRequest request,
        CancellationToken cancellationToken);
}

public sealed record ProviderPolicyAuthorityRotationDistributionResult(
    string Code,
    long Generation,
    ProviderPublisherTrustPolicyAuthorityId ActiveAuthority,
    ProviderPolicyAuthorityFloor Floor)
{
    public bool IsApplied => Code == "policy-authority-distribution-applied";
}

public static class ProviderPolicyAuthorityRotationDistributionManifest
{
    public static byte[] Encode(ProviderPolicyAuthorityRotationDistributionResponse response)
    {
        ArgumentNullException.ThrowIfNull(response);
        using var output = new MemoryStream();
        Append(output, "CBI58");
        Append(output, response.Challenge);
        Append(output, response.PolicySequence);
        AppendOptional(output, response.PolicyIdentity);
        Append(output, response.AuthorityGeneration);
        Append(output, response.ActiveAuthority.Value);
        Append(output, response.IssuedAtUnixSeconds);
        Append(output, response.ExpiresAtUnixSeconds);
        Append(output, response.Rotation is null ? 0 : 1);
        if (response.Rotation is not null) Append(output, RotationDigest(response.Rotation));
        return output.ToArray();
    }

    public static string RotationDigest(ProviderPolicyAuthorityRotationStatement rotation)
    {
        ArgumentNullException.ThrowIfNull(rotation);
        using var output = new MemoryStream();
        Append(output, "CBI58-ROTATION");
        Append(output, rotation.Generation);
        Append(output, rotation.PolicySequence);
        AppendOptional(output, rotation.PolicyIdentity);
        Append(output, rotation.PreviousAuthority.Value);
        Append(output, rotation.NextAuthority.Value);
        Append(output, rotation.Algorithm);
        Append(output, rotation.PreviousAuthorityPublicKeySpkiBase64);
        Append(output, rotation.NextAuthorityPublicKeySpkiBase64);
        Append(output, rotation.PreviousSignatureBase64);
        Append(output, rotation.NextSignatureBase64);
        return Convert.ToHexString(SHA256.HashData(output.ToArray()));
    }

    private static void AppendOptional(Stream output, ProviderPublisherTrustPolicyId? identity)
    {
        Append(output, identity.HasValue ? 1 : 0);
        if (identity.HasValue) Append(output, identity.Value.Value);
    }

    private static void Append(Stream output, int value)
    {
        Span<byte> bytes = stackalloc byte[4];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, long value)
    {
        Span<byte> bytes = stackalloc byte[8];
        BinaryPrimitives.WriteInt64BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, string value)
    {
        var bytes = new UTF8Encoding(false, true).GetBytes(value);
        Append(output, bytes.Length);
        output.Write(bytes);
    }
}

public sealed class ProviderPolicyAuthorityRotationDistributionClient
{
    private const int MaximumCharacters = 1024 * 1024;
    private static readonly TimeSpan MaximumValidity = TimeSpan.FromMinutes(15);
    private static readonly TimeSpan MaximumFutureSkew = TimeSpan.FromSeconds(30);
    private readonly DurableProviderPublisherTrustPolicyRegistry registry;
    private readonly ProviderPublisherTrustPolicyDistributionEndpointId endpointIdentity;

    public ProviderPolicyAuthorityRotationDistributionClient(
        DurableProviderPublisherTrustPolicyRegistry registry,
        ProviderPublisherTrustPolicyDistributionEndpointId endpointIdentity)
    {
        ArgumentNullException.ThrowIfNull(registry);
        if (endpointIdentity.Value is null) throw new ArgumentException("An endpoint identity is required.", nameof(endpointIdentity));
        this.registry = registry;
        this.endpointIdentity = endpointIdentity;
    }

    public async Task<ProviderPolicyAuthorityRotationDistributionResult> SynchronizeAsync(
        IProviderPolicyAuthorityRotationDistributionSource source,
        DateTimeOffset now,
        TimeSpan timeout,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(source);
        if (timeout <= TimeSpan.Zero || timeout > TimeSpan.FromMinutes(1))
            throw new ArgumentOutOfRangeException(nameof(timeout));
        var current = registry.Current;
        var request = new ProviderPolicyAuthorityRotationDistributionRequest(
            Convert.ToHexString(RandomNumberGenerator.GetBytes(32)), current?.Sequence ?? 0,
            current?.Policy.Identity, registry.AuthorityGeneration, registry.ActiveAuthorityIdentity);
        ProviderPolicyAuthorityRotationDistributionResponse response;
        using var timeoutSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutSource.CancelAfter(timeout);
        try
        {
            response = await source.FetchAsync(request, timeoutSource.Token)
                .WaitAsync(timeout, cancellationToken).ConfigureAwait(false);
        }
        catch (TimeoutException) { return Result("policy-authority-distribution-timeout"); }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        { return Result("policy-authority-distribution-canceled"); }
        catch (OperationCanceledException) { return Result("policy-authority-distribution-timeout"); }
        catch (Exception exception) when (exception is not OutOfMemoryException)
        { return Result("policy-authority-distribution-transport-failed"); }

        if (!IsBounded(response)) return Result("policy-authority-distribution-response-invalid");
        try
        {
            var publicKey = Convert.FromBase64String(response.EndpointPublicKeySpkiBase64);
            var supplied = ProviderPublisherTrustPolicyDistributionEndpointId.Create(
                Convert.ToHexString(SHA256.HashData(publicKey)));
            if (supplied != endpointIdentity) return Result("policy-authority-distribution-endpoint-mismatch");
            using var verifier = ECDsa.Create();
            verifier.ImportSubjectPublicKeyInfo(publicKey, out var bytesRead);
            var parameters = verifier.ExportParameters(false);
            if (bytesRead != publicKey.Length || verifier.KeySize != 256
                || parameters.Curve.Oid.Value != "1.2.840.10045.3.1.7")
                return Result("policy-authority-distribution-response-invalid");
            var signature = Convert.FromBase64String(response.SignatureBase64);
            if (!verifier.VerifyData(ProviderPolicyAuthorityRotationDistributionManifest.Encode(response),
                signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence))
                return Result("policy-authority-distribution-signature-invalid");
        }
        catch (Exception exception) when (exception is FormatException or CryptographicException
            or ArgumentException or NullReferenceException)
        { return Result("policy-authority-distribution-response-invalid"); }

        if (response.Challenge != request.Challenge)
            return Result("policy-authority-distribution-challenge-mismatch");
        if (response.PolicySequence != request.PolicySequence || response.PolicyIdentity != request.PolicyIdentity
            || response.AuthorityGeneration != request.AuthorityGeneration
            || response.ActiveAuthority != request.ActiveAuthority)
            return Result("policy-authority-distribution-cursor-mismatch");
        try
        {
            var issued = DateTimeOffset.FromUnixTimeSeconds(response.IssuedAtUnixSeconds);
            var expires = DateTimeOffset.FromUnixTimeSeconds(response.ExpiresAtUnixSeconds);
            if ((issued > now && issued - now > MaximumFutureSkew) || expires <= now
                || expires <= issued || expires - issued > MaximumValidity)
                return Result("policy-authority-distribution-stale");
        }
        catch (ArgumentOutOfRangeException) { return Result("policy-authority-distribution-stale"); }
        if (!CursorMatches(request)) return Result("policy-authority-distribution-superseded");
        if (response.Rotation is null) return Result("policy-authority-distribution-current");
        var applied = registry.Rotate(response.Rotation);
        return applied.IsApplied
            ? Result("policy-authority-distribution-applied")
            : Result(applied.Code);
    }

    private bool CursorMatches(ProviderPolicyAuthorityRotationDistributionRequest request)
    {
        var current = registry.Current;
        return request.PolicySequence == (current?.Sequence ?? 0)
            && request.PolicyIdentity == current?.Policy.Identity
            && request.AuthorityGeneration == registry.AuthorityGeneration
            && request.ActiveAuthority == registry.ActiveAuthorityIdentity;
    }

    private ProviderPolicyAuthorityRotationDistributionResult Result(string code) =>
        new(code, registry.AuthorityGeneration, registry.ActiveAuthorityIdentity, registry.AuthorityFloor);

    private static bool IsBounded(ProviderPolicyAuthorityRotationDistributionResponse? response)
    {
        if (response is null || response.Challenge is null || response.Algorithm != "ECDSA-P256-SHA256"
            || response.EndpointPublicKeySpkiBase64 is null || response.SignatureBase64 is null
            || response.ActiveAuthority.Value is null || response.Challenge.Length != 64
            || response.Challenge.Any(character => !Uri.IsHexDigit(character))) return false;
        long count = response.Challenge.Length + response.Algorithm.Length
            + response.EndpointPublicKeySpkiBase64.Length + response.SignatureBase64.Length
            + response.ActiveAuthority.Value.Length;
        if (response.Rotation is not null)
        {
            var rotation = response.Rotation;
            if (rotation.Algorithm is null || rotation.PreviousAuthorityPublicKeySpkiBase64 is null
                || rotation.NextAuthorityPublicKeySpkiBase64 is null || rotation.PreviousSignatureBase64 is null
                || rotation.NextSignatureBase64 is null) return false;
            count += rotation.Algorithm.Length + rotation.PreviousAuthorityPublicKeySpkiBase64.Length
                + rotation.NextAuthorityPublicKeySpkiBase64.Length + rotation.PreviousSignatureBase64.Length
                + rotation.NextSignatureBase64.Length;
        }
        return count <= MaximumCharacters;
    }
}

File: Minimal/src/Brontide.Minimal.Host/ProviderPolicyAuthorityRotationDistribution.fs
namespace Brontide.Minimal.Host

open System
open System.Buffers.Binary
open System.IO
open System.Security.Cryptography
open System.Text
open System.Threading
open System.Threading.Tasks

type ProviderPolicyAuthorityRotationDistributionRequest =
    { Challenge: string
      PolicySequence: int64
      PolicyIdentity: ProviderPublisherTrustPolicyId option
      AuthorityGeneration: int64
      ActiveAuthority: ProviderPublisherTrustPolicyAuthorityId }

type ProviderPolicyAuthorityRotationDistributionResponse =
    { Challenge: string
      PolicySequence: int64
      PolicyIdentity: ProviderPublisherTrustPolicyId option
      AuthorityGeneration: int64
      ActiveAuthority: ProviderPublisherTrustPolicyAuthorityId
      IssuedAtUnixSeconds: int64
      ExpiresAtUnixSeconds: int64
      Rotation: ProviderPolicyAuthorityRotationStatement option
      Algorithm: string
      EndpointPublicKeySpkiBase64: string
      SignatureBase64: string }

type IProviderPolicyAuthorityRotationDistributionSource =
    abstract FetchAsync:
        ProviderPolicyAuthorityRotationDistributionRequest * CancellationToken ->
            Task<ProviderPolicyAuthorityRotationDistributionResponse>

type ProviderPolicyAuthorityRotationDistributionResult =
    { Code: string
      Generation: int64
      ActiveAuthority: ProviderPublisherTrustPolicyAuthorityId
      Floor: ProviderPolicyAuthorityFloor }
    member this.IsApplied = this.Code = "policy-authority-distribution-applied"

[<RequireQualifiedAccess>]
module ProviderPolicyAuthorityRotationDistributionManifest =
    let private appendInt32 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> 4
        BinaryPrimitives.WriteInt32BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendInt64 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> 8
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendString output (value: string) =
        let bytes = UTF8Encoding(false, true).GetBytes value
        appendInt32 output bytes.Length
        output.Write bytes
    let private appendOptional output identity =
        appendInt32 output (if Option.isSome identity then 1 else 0)
        identity |> Option.iter (ProviderPublisherTrustPolicyId.value >> appendString output)

    let rotationDigest (rotation: ProviderPolicyAuthorityRotationStatement) =
        if isNull (box rotation) then nullArg (nameof rotation)
        use output = new MemoryStream()
        appendString output "CBI58-ROTATION"
        appendInt64 output rotation.Generation
        appendInt64 output rotation.PolicySequence
        appendOptional output rotation.PolicyIdentity
        appendString output (ProviderPublisherTrustPolicyAuthorityId.value rotation.PreviousAuthority)
        appendString output (ProviderPublisherTrustPolicyAuthorityId.value rotation.NextAuthority)
        appendString output rotation.Algorithm
        appendString output rotation.PreviousAuthorityPublicKeySpkiBase64
        appendString output rotation.NextAuthorityPublicKeySpkiBase64
        appendString output rotation.PreviousSignatureBase64
        appendString output rotation.NextSignatureBase64
        output.ToArray() |> SHA256.HashData |> Convert.ToHexString

    let encode (response: ProviderPolicyAuthorityRotationDistributionResponse) =
        if isNull (box response) then nullArg (nameof response)
        use output = new MemoryStream()
        appendString output "CBI58"
        appendString output response.Challenge
        appendInt64 output response.PolicySequence
        appendOptional output response.PolicyIdentity
        appendInt64 output response.AuthorityGeneration
        appendString output (ProviderPublisherTrustPolicyAuthorityId.value response.ActiveAuthority)
        appendInt64 output response.IssuedAtUnixSeconds
        appendInt64 output response.ExpiresAtUnixSeconds
        appendInt32 output (if Option.isSome response.Rotation then 1 else 0)
        response.Rotation |> Option.iter (rotationDigest >> appendString output)
        output.ToArray()

type ProviderPolicyAuthorityRotationDistributionClient(
    registry: DurableProviderPublisherTrustPolicyRegistry,
    endpointIdentity: ProviderPublisherTrustPolicyDistributionEndpointId) =

    do
        if isNull (box registry) then nullArg (nameof registry)
        if isNull (box endpointIdentity) then invalidArg (nameof endpointIdentity) "An endpoint identity is required."

    let maximumCharacters = 1024L * 1024L
    let maximumValidity = TimeSpan.FromMinutes 15.0
    let maximumFutureSkew = TimeSpan.FromSeconds 30.0

    let result code =
        { Code = code
          Generation = registry.AuthorityGeneration
          ActiveAuthority = registry.ActiveAuthorityIdentity
          Floor = registry.AuthorityFloor }

    let cursorMatches (request: ProviderPolicyAuthorityRotationDistributionRequest) =
        let current = registry.Current
        request.PolicySequence = (current |> Option.map _.Sequence |> Option.defaultValue 0L)
        && request.PolicyIdentity = (current |> Option.map _.Policy.Identity)
        && request.AuthorityGeneration = registry.AuthorityGeneration
        && request.ActiveAuthority = registry.ActiveAuthorityIdentity

    let bounded (response: ProviderPolicyAuthorityRotationDistributionResponse) =
        if isNull (box response) || String.IsNullOrEmpty response.Challenge || response.Algorithm <> "ECDSA-P256-SHA256"
            || String.IsNullOrEmpty response.EndpointPublicKeySpkiBase64 || String.IsNullOrEmpty response.SignatureBase64
            || isNull (box response.ActiveAuthority)
            || response.Challenge.Length <> 64 || response.Challenge |> Seq.exists (Uri.IsHexDigit >> not) then false
        else
            let mutable count = int64 (response.Challenge.Length + response.Algorithm.Length
                + response.EndpointPublicKeySpkiBase64.Length + response.SignatureBase64.Length
                + (ProviderPublisherTrustPolicyAuthorityId.value response.ActiveAuthority).Length)
            match response.Rotation with
            | Some rotation when String.IsNullOrEmpty rotation.Algorithm
                || String.IsNullOrEmpty rotation.PreviousAuthorityPublicKeySpkiBase64
                || String.IsNullOrEmpty rotation.NextAuthorityPublicKeySpkiBase64
                || String.IsNullOrEmpty rotation.PreviousSignatureBase64
                || String.IsNullOrEmpty rotation.NextSignatureBase64 -> false
            | Some rotation ->
                count <- count + int64 (rotation.Algorithm.Length + rotation.PreviousAuthorityPublicKeySpkiBase64.Length
                    + rotation.NextAuthorityPublicKeySpkiBase64.Length + rotation.PreviousSignatureBase64.Length
                    + rotation.NextSignatureBase64.Length)
                count <= maximumCharacters
            | None -> count <= maximumCharacters

    member _.SynchronizeAsync(
        source: IProviderPolicyAuthorityRotationDistributionSource,
        now: DateTimeOffset,
        timeout: TimeSpan,
        cancellationToken: CancellationToken) = task {
        if isNull (box source) then nullArg (nameof source)
        if timeout <= TimeSpan.Zero || timeout > TimeSpan.FromMinutes 1.0 then
            invalidArg (nameof timeout) "The timeout must be positive and no greater than one minute."
        let current = registry.Current
        let request =
            { Challenge = RandomNumberGenerator.GetBytes 32 |> Convert.ToHexString
              PolicySequence = current |> Option.map _.Sequence |> Option.defaultValue 0L
              PolicyIdentity = current |> Option.map _.Policy.Identity
              AuthorityGeneration = registry.AuthorityGeneration
              ActiveAuthority = registry.ActiveAuthorityIdentity }
        use timeoutSource = CancellationTokenSource.CreateLinkedTokenSource cancellationToken
        timeoutSource.CancelAfter timeout
        let! fetched = task {
            try
                let! response = source.FetchAsync(request, timeoutSource.Token).WaitAsync(timeout, cancellationToken)
                return Ok response
            with
            | :? TimeoutException -> return Error "policy-authority-distribution-timeout"
            | :? OperationCanceledException when cancellationToken.IsCancellationRequested ->
                return Error "policy-authority-distribution-canceled"
            | :? OperationCanceledException -> return Error "policy-authority-distribution-timeout"
            | error when not (error :? OutOfMemoryException) ->
                return Error "policy-authority-distribution-transport-failed"
        }
        match fetched with
        | Error code -> return result code
        | Ok response when not (bounded response) ->
            return result "policy-authority-distribution-response-invalid"
        | Ok response ->
            let verification =
                try
                    let publicKey = Convert.FromBase64String response.EndpointPublicKeySpkiBase64
                    let supplied = publicKey |> SHA256.HashData |> Convert.ToHexString
                                   |> ProviderPublisherTrustPolicyDistributionEndpointId.create
                    if supplied <> endpointIdentity then Error "policy-authority-distribution-endpoint-mismatch"
                    else
                        use verifier = ECDsa.Create()
                        let mutable bytesRead = 0
                        verifier.ImportSubjectPublicKeyInfo(publicKey.AsSpan(), &bytesRead)
                        let curve = verifier.ExportParameters(false).Curve.Oid.Value |> Option.ofObj
                        if bytesRead <> publicKey.Length || verifier.KeySize <> 256
                            || curve <> Some "1.2.840.10045.3.1.7" then
                            Error "policy-authority-distribution-response-invalid"
                        else
                            let signature = Convert.FromBase64String response.SignatureBase64
                            if verifier.VerifyData(
                                ProviderPolicyAuthorityRotationDistributionManifest.encode response,
                                signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence) then Ok ()
                            else Error "policy-authority-distribution-signature-invalid"
                with
                | :? FormatException
                | :? CryptographicException
                | :? ArgumentException
                | :? NullReferenceException -> Error "policy-authority-distribution-response-invalid"
            match verification with
            | Error code -> return result code
            | Ok () when response.Challenge <> request.Challenge ->
                return result "policy-authority-distribution-challenge-mismatch"
            | Ok () when response.PolicySequence <> request.PolicySequence
                || response.PolicyIdentity <> request.PolicyIdentity
                || response.AuthorityGeneration <> request.AuthorityGeneration
                || response.ActiveAuthority <> request.ActiveAuthority ->
                return result "policy-authority-distribution-cursor-mismatch"
            | Ok () ->
                let fresh =
                    try
                        let issued = DateTimeOffset.FromUnixTimeSeconds response.IssuedAtUnixSeconds
                        let expires = DateTimeOffset.FromUnixTimeSeconds response.ExpiresAtUnixSeconds
                        (issued <= now || issued - now <= maximumFutureSkew) && expires > now && expires > issued
                        && expires - issued <= maximumValidity
                    with :? ArgumentOutOfRangeException -> false
                if not fresh then return result "policy-authority-distribution-stale"
                elif not (cursorMatches request) then return result "policy-authority-distribution-superseded"
                else
                    match response.Rotation with
                    | None -> return result "policy-authority-distribution-current"
                    | Some rotation ->
                        let applied = registry.Rotate rotation
                        return result (if applied.IsApplied then "policy-authority-distribution-applied" else applied.Code)
    }
